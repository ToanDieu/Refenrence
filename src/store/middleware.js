import { matchPath } from "react-router-dom";
import { CHANGE_PAGE, NOTIFY_BUTTON_CLICK } from "../constants/actionTypes";

const changeToBasePage = id => ({
  type: CHANGE_PAGE,
  payload: {
    pageType: "base",
    detail: { id }
  }
});

export const trackPageNavigate = ({
  dispatch
  // getState
}) => next => action => {
  if (action.type === "@@router/LOCATION_CHANGE") {
    console.log("trackViews__action: ", action);
    const pathName = action.payload.location.pathname;

    const isEnteringBaseCases = matchPath(pathName, {
      path: "/bases/:baseID/cases"
    });

    const isEnteringBaseDetail = matchPath(pathName, {
      path: "/bases/:baseID"
    });

    // console.log("isEnteringBaseDetail", isEnteringBaseDetail, action);

    const isEnteringBaseWorkflow = matchPath(pathName, {
      path: "/bases/:baseID/workflows/*/actions"
    });

    const isEnteringNotification = matchPath(pathName, {
      path: "/notifications"
    });

    if (isEnteringNotification) {
      dispatch({
        type: NOTIFY_BUTTON_CLICK
      });
    }

    if (isEnteringBaseCases) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseCases.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseDetail) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseDetail.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseWorkflow) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseWorkflow.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }
  }

  return next(action);
};
