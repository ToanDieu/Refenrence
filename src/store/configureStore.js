import { createStore, compose, applyMiddleware } from "redux";
import reduxImmutableStateInvariant from "redux-immutable-state-invariant";
import thunk from "redux-thunk";
import { routerMiddleware } from "connected-react-router";

import history from "@/utils/history";
import createReducer from "../reducers";
import { trackPageNavigate } from "./middleware";

function configureStoreProd(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [thunk, reactRouterMiddleware, trackPageNavigate];

  const store = createStore(
    createReducer(),
    initialState,
    compose(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  return store;
}

function configureStoreDev(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [
    reduxImmutableStateInvariant(),
    thunk,
    reactRouterMiddleware,
    trackPageNavigate
  ];

  /* eslint-disable no-underscore-dangle */
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; // add support for Redux dev tools
  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept("../reducers", () => {
      const nextReducer = require("../reducers").default; // eslint-disable-line global-require
      store.replaceReducer(nextReducer);
    });
  }

  return store;
}

const store =
  process.env.NODE_ENV === "production"
    ? configureStoreProd()
    : configureStoreDev();

export { history, store };
