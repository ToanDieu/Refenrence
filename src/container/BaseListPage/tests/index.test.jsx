import React from "react";
import { render } from "react-testing-library";
import { Provider } from "react-redux";
import { initialize, addTranslationForLanguage } from "react-localize-redux";

import languages from "@/i18n/languages.json";
import english from "@/i18n/en.json";
import { store as appStore } from "@/store/configureStore";
import { BaseListPage, mapDispatchToProps } from "../index";

describe("<BaseListPage />", () => {
  let store;

  beforeAll(() => {
    store = appStore;
    const defaultLanguage = "en";
    store.dispatch(initialize(languages, { defaultLanguage }));
    store.dispatch(addTranslationForLanguage(english, "en"));
  });

  it("should render and match the snapshot", () => {
    const { container } = render(
      <Provider store={store}>
        <BaseListPage bases={[]} translate={jest.fn()} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render and match the snapshot on loading", () => {
    const { container } = render(
      <Provider store={store}>
        <BaseListPage loading bases={[]} translate={jest.fn()} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should fetch data on mount", () => {
    const spy = jest.fn();
    render(
      <Provider store={store}>
        <BaseListPage
          typeID={1}
          bases={[]}
          fetchAll={spy}
          translate={jest.fn()}
        />
      </Provider>
    );
    expect(spy).toHaveBeenCalled();
  });

  describe("mapDispatchToProps", () => {
    describe("fetchAll", () => {
      it("should be injected", () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.fetchAll).toBeDefined();
      });

      it("should dispatch fetchAll when called", () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const typeID = 1;
        result.fetchAll(typeID);
        expect(dispatch).toHaveBeenCalled();
      });
    });
  });
});
