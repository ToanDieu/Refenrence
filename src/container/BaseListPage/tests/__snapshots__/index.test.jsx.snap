// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseListPage /> should render and match the snapshot 1`] = `
.emotion-3 {
  padding: 47px 20px 93px 20px;
  max-width: 1245px;
  margin: auto;
  min-height: calc(100vh - 156px);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 20px;
  font-weight: 500;
  text-transform: uppercase;
  line-height: 1.5;
  margin-bottom: 22px;
}

.emotion-2 {
  background-color: #fff;
  border-radius: 2px;
  text-align: left;
  font-size: 15px;
  width: 100%;
}

.emotion-2 thead th {
  border-bottom: 1px solid #e8e6e2;
  padding: 26px 14px 19px;
  font-weight: 500;
  text-transform: uppercase;
  line-height: 1.2;
}

.emotion-2 tbody tr:nth-child(odd) {
  background-color: #fafaf9;
}

.emotion-2 td {
  padding: 18px;
  vertical-align: middle;
}

.emotion-2 td:nth-child(1) {
  width: 250px;
}

.emotion-2 td:nth-child(2) {
  width: 100px;
}

.emotion-2 td:nth-child(3) {
  width: 220px;
}

.emotion-2 td:nth-child(5),
.emotion-2 td:nth-child(6) {
  padding-left: 0;
  width: 24px;
}

<article
  class="emotion-3"
>
  <div
    class="emotion-1"
  >
    <h1
      class="emotion-0"
    >
      Bases
    </h1>
    <div>
      <div
        class="el-dropdown"
      >
        <span
          class="el-dropdown-link"
        >
          <img
            class="icon-img--24 el-icon-caret-bottom el-icon--right"
            src="IMAGE_MOCK"
          />
        </span>
        <ul
          class="el-dropdown-menu"
          style="display: none;"
        >
          <li
            class="el-dropdown-menu__item"
          >
            <div
              class=""
            >
              All
            </div>
          </li>
          <div
            class=""
          />
        </ul>
      </div>
    </div>
  </div>
  <table
    class="emotion-2"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Code
        </th>
        <th>
          Updated
        </th>
        <th />
        <th />
        <th />
      </tr>
    </thead>
    <tbody />
  </table>
</article>
`;

exports[`<BaseListPage /> should render and match the snapshot on loading 1`] = `
.emotion-2 {
  padding: 47px 20px 93px 20px;
  max-width: 1245px;
  margin: auto;
  min-height: calc(100vh - 156px);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 20px;
  font-weight: 500;
  text-transform: uppercase;
  line-height: 1.5;
  margin-bottom: 22px;
}

<article
  class="emotion-2"
>
  <div
    class="emotion-1"
  >
    <h1
      class="emotion-0"
    >
      Bases
    </h1>
    <div>
      <div
        class="el-dropdown"
      >
        <span
          class="el-dropdown-link"
        >
          <img
            class="icon-img--24 el-icon-caret-bottom el-icon--right"
            src="IMAGE_MOCK"
          />
        </span>
        <ul
          class="el-dropdown-menu"
          style="display: none;"
        >
          <li
            class="el-dropdown-menu__item"
          >
            <div
              class=""
            >
              All
            </div>
          </li>
          <div
            class=""
          />
        </ul>
      </div>
    </div>
  </div>
  <div
    class="spinner-wrapper"
  >
    <div
      class="spinner__dots spinner--row"
    >
      <div
        class="bounce1"
      />
      <div
        class="bounce2"
      />
      <div
        class="bounce3"
      />
    </div>
  </div>
</article>
`;
